### DISCUSSION TITLE
useFieldArray: removing element makes all other elements dirty

### DISCUSSION THREAD

**Discussion by k0l11** [react-hook-form/react-hook-form#9634] [Q&A]
Hi, thanks for the great library! Currently I'm trying to use useFieldArray together with provided defaultValues and dirtyFields to display what fields of a specific element the user has changed in the form. This works fine as long as I don't delete an element, but the moment I delete one of the elements via `remove()` all of the remaining elements have their inputs set to dirty in dirtyFields. 
Is this intended? What's the recommended way to work around this? Instead of removing the element I could set a flag in the element itself via `update()` to remove it from the displayed elements, but this feels hacky to me.

**Comment by leapful** [comment]
@k0l11 

When working with field array, order is matter and depends on `defaultValues`.

For example:

```
// Giving default values

{ 
   test: [
       { name: 'A' },
       { name: 'B' },
       { name: 'C' },
   ]
}

// Then we will have a list with order

A = index 0
B = index 1
C = index 2

// When removing B

A = index 0
C = index 1

```

=> Comparing with default values, we have B and C are all dirty

If you want to change the dirty behavior checking, just update default values for field array after removing one item by using `reset` method.

Working demo: https://codesandbox.io/s/staging-voice-m8uk9i?file=/src/App.tsx

**Reply by k0l11** [reply]
Thanks for the answer, but calling reset(...) on the form like this doesn't preserve the "dirtyness" of A and C which is something I want.

**Comment by Mario-Eis** [comment]
The best you can do is to avoid useFieldArray completely. Unfortunately it breaks the dirty state. And there is no plan to change that. See https://github.com/react-hook-form/react-hook-form/issues/11402.
We had to use a flag in the forms data to mark entries as "deleted". And use setValue to change values and add values to the array. So yes, the hacky way you suggested is the way to go. useFieldArray is unusable without proper dirty field handling.
